### ISSUE TITLE
Clarify what event is `userEvent` to consider `click`/`auxclick` event target after `touchend` (or `pointerup`)

### ISSUE THREAD

**Issue by masayuki-nakano** [w3c/pointerevents#508]
Pointer Events [defines](https://w3c.github.io/pointerevents/#event-dispatch) that `click`/`auxclick` event target is:

> 2. If userEvent is not a PointerEvent, dispatch event following the [[UIEVENTS](https://w3c.github.io/pointerevents/#bib-uievents)] spec without overriding event target and skip the remaining steps below.
> 3. If event is a contextmenu event, or userEvent was dispatched while the corresponding pointer was captured, then let target be the target of userEvent.

and `userEvent` is defined as:

> userEvent be the user interaction event that caused the firing of event.

If you touch the device with a single input, `touchstart` and `touchend` may cause compatibility mouse events, `mousemove`, `mousedown` and `mouseup`. However, web apps consumes `pointerdown` or `touchstart`, the compatibility mouse events are not fired. Therefore, `userEvent` may be `pointerup`, `touchend` or `mouseup`. The compatibility mouse event target is [defined](https://w3c.github.io/touch-events/#mouse-events) as the element at the `touchend` position. The `touchend` event target is [defined](https://w3c.github.io/touch-events/#event-touchend) as the element which is the target of `touchstart`. So, these events may be targeted different elements. Therefore, `click`/`auxclick` event target may be different from which event is `userEvent`.

**Comment by masayuki-nakano** [comment]
Oh, if `touchstart` is canceled, `click` should not be fired.

**Comment by masayuki-nakano** [comment]
Created [a new test](https://github.com/web-platform-tests/wpt/pull/46805) which (probably) conforms to current event specs. However, as I mentioned in the commit message, conforming to the test result may break bxSlider.

**Comment by masayuki-nakano** [comment]
Ccing: @mustaqahmed, @smaug----, @flackr

**Comment by masayuki-nakano** [comment]
(FYI: This issue blocks Mozilla fixing the failures of `pointerevent_click_during_capture.html`.)

**Comment by masayuki-nakano** [comment]
Currently, if I change Firefox behavior to conform to the spec, [links in this bxSlider deom](https://bxslider.com/examples) works, but the originally reported example has gone. Therefore, I don't know whether bxSlider has fixed [the issue](https://github.com/stevenwanderski/bxslider-4/issues/1188) actually.

**Comment by masayuki-nakano** [comment]
Sorting out the issues:
* `click` event target may be different even if same position is clicked/tapped, depending on `pointerType` is `mouse` or `touch` (see the comments in my testcase)
* some developers might expect that capturing parent event target of `pointerdown` won't stop dispatching `click` on the original target (but the reported web site seems working with current defined behavior today)

At least the former breaks a Pointer Event purpose to handle all pointing device inputs with one set of DOM events.

**Comment by flackr** [comment]
The note in that section implies (at least to me) that the userEvent is the PointerEvent (i.e. pointerup) whose default action caused the dispatch (excepting keyboard / accessibility means of dispatching clicks):

> Event userEvent could be a non-PointerEvent; for example, it is a KeyboardEvent when a click event dispatch is caused by hitting the spacebar on a checkbox element.
>
> When userEvent is a PointerEvent, userEvent is a [pointerup](https://w3c.github.io/pointerevents/#dfn-pointerup) for a click or auxclick event, and either a [pointerdown](https://w3c.github.io/pointerevents/#dfn-pointerdown) or a [pointerup](https://w3c.github.io/pointerevents/#dfn-pointerup) event (depending on native platform convention) for a contextmenu event.

But I'm not sure I understand the specific differences that are causing problems. Most pages won't be setting pointerCapture so the target of the click event will be the same as it usually is.

**Comment by masayuki-nakano** [comment]
> The note in that section implies (at least to me) that the userEvent is the PointerEvent (i.e. pointerup) whose default action caused the dispatch (excepting keyboard / accessibility means of dispatching clicks):

If so, the `click` event target won't depend on `mouse` or `touch`, but both Chrome and Firefox dispatch it on wrong target, they dispatch `click` on the preceding `touchend` target which is same as `touchstart`'s.

**Comment by flackr** [comment]
> > The note in that section implies (at least to me) that the userEvent is the PointerEvent (i.e. pointerup) whose default action caused the dispatch (excepting keyboard / accessibility means of dispatching clicks):
> 
> If so, the `click` event target won't depend on `mouse` or `touch`, but both Chrome and Firefox dispatch it on wrong target, they dispatch `click` on the preceding `touchend` target which is same as `touchstart`'s.

Is this not the same as the pointerup target? Note that touch usually has [implicit capture](https://w3c.github.io/pointerevents/#implicit-pointer-capture) so these would all be the same target AFAIK.

**Comment by masayuki-nakano** [comment]
> > > The note in that section implies (at least to me) that the userEvent is the PointerEvent (i.e. pointerup) whose default action caused the dispatch (excepting keyboard / accessibility means of dispatching clicks):
> > 
> > 
> > If so, the `click` event target won't depend on `mouse` or `touch`, but both Chrome and Firefox dispatch it on wrong target, they dispatch `click` on the preceding `touchend` target which is same as `touchstart`'s.
> 
> Is this not the same as the pointerup target? Note that touch usually has [implicit capture](https://w3c.github.io/pointerevents/#implicit-pointer-capture) so these would all be the same target AFAIK.

No if `pointerdown` listener makes different element from `pointerdown`/`touchstart`/`touchend` target capture the pointer. [The expected result of line 192 of my test](https://github.com/web-platform-tests/wpt/blob/f18377726f412c270c3e3ed98c0b9f1c65ec2eac/pointerevents/pointerevent_click_during_parent_capture.html#L192) passed on Chrome and Firefox.

**Comment by flackr** [comment]
> No if pointerdown listener makes different element from pointerdown/touchstart/touchend target capture the pointer. [The expected result of line 192 of my test](https://github.com/web-platform-tests/wpt/blob/f18377726f412c270c3e3ed98c0b9f1c65ec2eac/pointerevents/pointerevent_click_during_parent_capture.html#L192) passed on Chrome and Firefox.

If pointerdown captures the pointer to a different element, pointerup should "target" (via capture) that different element too though.

**Comment by masayuki-nakano** [comment]
> > No if pointerdown listener makes different element from pointerdown/touchstart/touchend target capture the pointer. [The expected result of line 192 of my test](https://github.com/web-platform-tests/wpt/blob/f18377726f412c270c3e3ed98c0b9f1c65ec2eac/pointerevents/pointerevent_click_during_parent_capture.html#L192) passed on Chrome and Firefox.
> 
> If pointerdown captures the pointer to a different element, pointerup should "target" (via capture) that different element too though.

Yeah, however, both Chrome and Firefox dispatch `click` event on the `#target` instead of `#parent` which is captured and is the target of `pointerup`. So, both browsers treat `userEvent` as the preceding `touchend` or `mouseup`.

**Comment by flackr** [comment]
> Yeah, however, both Chrome and Firefox dispatch `click` event on the `#target` instead of `#parent` which is captured and is the target of `pointerup`. So, both browsers treat `userEvent` as the preceding `touchend` or `mouseup`.

I believe if you try chrome with the `--enable-blink-features=BoundaryEventDispatchTracksNodeRemoval` flag that it will dispatch to the captured target of the pointerevent instead (code [here](https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/core/input/pointer_event_manager.cc;l=1091;bpv=1;bpt=1?q=pointereventmanager)). @mustaqahmed I'm not sure why we have this change behind that flag, I thought this was a separate issue.

**Comment by masayuki-nakano** [comment]
> I believe if you try chrome with the `--enable-blink-features=BoundaryEventDispatchTracksNodeRemoval` flag that it will dispatch to the captured target of the pointerevent instead (code [here](https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/core/input/pointer_event_manager.cc;l=1091;bpv=1;bpt=1?q=pointereventmanager)). 

Well, I assume that it's enabled by default when I run WPT on Chrome from `./mach wpt` command in Firefox development environment (I use Chrome Canary). The command does not accept to explicitly enable it, but in the same condition, all tests of [pointerevent_after_target_removed.html?mouse](https://wpt.fyi/results/pointerevents/pointerevent_after_target_removed.html%3Fmouse?label=experimental&label=master&aligned&view=interop&q=label%3Ainterop-2023-events) pass so I believe that the feature is enabled by default when I run Chrome Canary from the command. However, my test unexpectedly pass on Chrome Canary.

When I run Chrome Canary, I see the following warning in the terminal:
```
 0:14.97 pid:56536 [4464:22700:0623/115359.500:WARNING:chrome_browser_main_win.cc(712)] Command line too long for RegisterApplicationRestart:  --allow-pre-commit-input --autoplay-policy=no-user-gesture-required --disable-background-networking --disable-backgrounding-occluded-windows --disable-client-side-phishing-detection --disable-default-apps --disable-hang-monitor --disable-infobars --disable-popup-blocking --disable-prompt-on-repost --disable-sync --enable-blink-features=DisableAhemAntialias --enable-experimental-web-platform-features --enable-features=FedCmWithoutWellKnownEnforcement,SecurePaymentConfirmationBrowser,GenericSensorExtraClasses --enable-logging=stderr --headless=new --host-resolver-rules="MAP nonexistent.*.test ^NOTFOUND, MAP *.test 127.0.0.1, MAP *.test. 127.0.0.1" --ignore-certificate-errors-spki-list=sCJ8962Wxqgz44IKoPQLcDT7YRRAxO2w1iYIqpMYHhg=,0Rt4mT6SJXojEMHTnKnlJ/hBKMBcI4kteBlhR1eTTdk= --ip-address-space-overrides=127.0.0.1:8002=private,127.0.0.1:8003=public,127.0.0.1:8445=private,127.0.0.1:8446=public --log-level=0 --no-first-run --no-service-autorun --noerrdialogs --password-store=basic --remote-debugging-port=0 --short-reporting-delay --test-type=webdriver --use-fake-device-for-media-stream --use-fake-ui-for-digital-identity --use-fake-ui-for-fedcm --use-fake-ui-for-media-stream --use-mock-keychain --user-data-dir="C:\Temp\scoped_dir56536_1771468703" --webtransport-developer-mode --restore-last-session --restart
```

Doesn't `--enable-experimental-web-platform-features` enable the feature as well?

**Comment by flackr** [comment]
> Doesn't --enable-experimental-web-platform-features enable the feature as well?

Ah, yes, it does. That turns on all features with status: experimental in runtime_enabled_features.json5 which it turns out [BoundaryEventDispatchTracksNodeRemoval](https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/platform/runtime_enabled_features.json5?q=BoundaryEventDispatchTracksNodeRemoval%20file:runtime_enabled_features%5C.json5) is.

> However, my test unexpectedly pass on Chrome Canary.

@mustaqahmed could you look into this test case?

**Comment by mustaqahmed** [comment]
@masayuki-nakano I missed the discussion above!  To clarify, the Chrome flag `BoundaryEventDispatchTracksNodeRemoval` is not related to `click`-like events at all! This flag controls only the dispatch of boundary (`enter`, `leave`, `over` and `out`) events after a DOM modification (tested through the WPTs [/pointerevents/pointerevent_after_target*](https://wpt.fyi/results/pointerevents?label=master&label=experimental&aligned&q=%2Fpointerevents%2Fpointerevent_after_target)).

Regarding your original discussion around `userEvent`: Chrome has bugs with click-like event's targets with pointer-capture.  [This bug](https://issues.chromium.org/40851596) is for mouse pointers, and we are planning to fix this as part of Interop 2024.  The behavior is tested through the WPT [pointerevent_click_during_capture.html](https://wpt.fyi/results/pointerevents?label=master&label=experimental&aligned&q=pointerevents%2Fpointerevent_click_during_capture.html).  I suspect we may have similar bugs with touch-pointers.

Since the definition of `userEvent` is perhaps causing the confusion here, please consider sending a PR to the spec to clarify it.  Do we need to add a note to the single sentence defining this (Step 1 in 4.2.12.3)?  This is meant to be the "low-level" event that caused the firing of a "click"-like event, regardless of the target of either event.

**Comment by mustaqahmed** [comment]
Here in my conclusion about this issue, please correct me if I missed something:
- We still don't have any evidence that the behavior defined here (and tested in [1](https://wpt.fyi/results/pointerevents/pointerevent_click_during_capture.html%3Fmouse-auxclick?label=experimental&label=master&aligned&view=interop&q=label%3Ainterop-2023-events) and [2](https://wpt.fyi/results/pointerevents/pointerevent_click_during_capture.html%3Fmouse-click?label=experimental&label=master&aligned&view=interop&q=label%3Ainterop-2023-events)) breaks the Web.  In other words, the risk remains unknown.
- Exposing this behavior (which is a part of Interop 2024) seems to be the only way to find out any compat implications, and no browsers has tried it yet.

**Comment by smaug----** [comment]
@masayuki-nakano do you have link to the Gecko bug which was dealing with bxSlider.

**Comment by masayuki-nakano** [comment]
@smaug---- it's [bug 1447993](https://bugzilla.mozilla.org/show_bug.cgi?id=1447993).

**Comment by masayuki-nakano** [comment]
> * We still don't have any evidence that the behavior defined here (and tested in [1](https://wpt.fyi/results/pointerevents/pointerevent_click_during_capture.html%3Fmouse-auxclick?label=experimental&label=master&aligned&view=interop&q=label%3Ainterop-2023-events) and [2](https://wpt.fyi/results/pointerevents/pointerevent_click_during_capture.html%3Fmouse-click?label=experimental&label=master&aligned&view=interop&q=label%3Ainterop-2023-events)) breaks the Web.  In other words, the risk remains unknown.
>
> * Exposing this behavior (which is a part of Interop 2024) seems to be the only way to find out any compat implications, and no browsers has tried it yet.

I have a concern about usability of web apps if browsers change the event target of touch events and compatibility mouse events. I found [a tricky code](https://searchfox.org/mozilla-central/rev/b60bff00c0fe136f35e4d6f3316911c090a39f23/layout/base/PositionedEventTargeting.cpp#434-475) in Firefox for Android, that tries to adjust event target of the compatibility mouse events to a closest element around the touch point if there is:
* the cursor style is "pointer"
* an element which listens to mouse/touch/pointer event
* a form control
* a label having target
* an element whose role is clickable
* an editable element

So, Firefox for Android makes web apps possible to handle within high-precision coordinates with touch/pointer event listeners but in rough coordinates with mouse event listeners including `click`.

After enabling the standardized behavior on Firefox, Firefox for Android users need to touch more strictly than current builds. I guess that Mozilla can update the tricky code for the new behavior, though.

If the other browsers do similar thing for users, they will also have similar regression.

**Comment by mustaqahmed** [comment]
Chromium does a similar [TouchAdjustment](https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/core/page/touch_adjustment.cc;drc=2614c4fd91c5e9336c33dbcca82e9b8038e8ebff;l=556), and we find it independent from the `click` target change we are discussing here.  Let me explain...

Chromium's TouchAdjustment is done on "low-level" (`pointerdown` and `pointerup`) events only, and then `click` target is determined latter from the effective targets of those low-level events, w/o any further adjustments.  The `userEvent` variable in the spec would affect only the `click` target in this case, and this target will be overridden only in the case the pointer has been captured.  Without any pointer-capture by the JS, the adjustment would work as before; and with pointer-capture, the JS fully-determines target of the `click` (which I think correctly reflects the main motivation behind the `userEvent` wording in the spec).

So TouchAdjustment complications feel like an isolated problem than `click` targets, at least until the planned change hits the Web.

**Comment by masayuki-nakano** [comment]
@mustaqahmed  Thank you for the explanation!

Does the "low-level" event mean internal event to handle default actions? I.e., DOM events exposed to the web may have different targets?

> Define target as follows:
>
> If event is a contextmenu event, or userEvent was dispatched while the corresponding pointer was captured, then let target be the target of userEvent.
>
> Otherwise **(event is a click or auxclick event for which userEvent is a pointerup event that was dispatched uncaptured) let target be the nearest common inclusive ancestor of the corresponding pointerdown and pointerup targets** in the DOM at the moment event is being dispatched.

Currently, Gecko uses same event instance to dispatch into the DOM and to handle default actions. Therefore, adjusting the pointer event target also changes the default action target. (Yeah, if current spec will become stable, Gecko needs to store the adjusted targets separately.)

**Comment by flackr** [comment]
I put together a demo to play with this, which shows indeed that chrome adjusts the target of all of the events where firefox and safari only adjust the compatibility events. https://flackr.github.io/web-demos/pointer-events/touch-adjustment/index.html

**Comment by masayuki-nakano** [comment]
Well, I was trying to sort out the adjustment issue, but it seems complicated.

However, the root cause is, the spec does not assume that all browsers adjust touch event targets and/or mouse event targets are adjusted to different targets.

I think that aligning the target to `pointerdown` and `pointerup` is too risky for touch devices from usability/accessibility points of view. Especially when the `pointerdown` event is not consumed. I believe that if mouse events are not dispatched by consuming `poitnerdown` ~~, `touchstart`, `touchmove` or `touchend`~~, etc, aligning the target must make sense since the web app is now aware of the target. On the other hand, if mouse events are fired, keeping the `click` related event targets as same as UI Events guarantees the backward compatibility especially web apps which handles only mouse events.

I'd like to suggest that the `click` event target alignment should be limited into less cases.

**Comment by masayuki-nakano** [comment]
Err, if `touchstart`, `touchmove` or `touchend` is canceled, `click` should not be fired too.

So, I think that in these cases, it's safe to not align the `click` event target to `pointerdown` and `pointerup` except when:
* explicitly an element captures the pointer at dispatching `pointerup`
* the preceding `pointerdown` was consumed and no mouse events are fired
So, I mean that it's not safe to change the `click` target if there is only implicit pointer capture at dispatching `pointerup`.

What do you think?

**Comment by mustaqahmed** [comment]
> So, I think that in these cases, it's safe to not align the click event target to pointerdown and pointerup except when...

We all agree there is a compat risk here (FYI, see the [Chrome bug](https://issues.chromium.org/40851596)).  Let me clarify why I am hesitant to change the spec-ed behavior now: we in PEWG have discussed click targets for captured pointers for several years (starting with #75 on 2016, then forked the discussion to #356 to focus more).  We even talked about risks w/o known evidence.  The spec text we are discussing here was updated only after [this conclusion](https://github.com/w3c/pointerevents/issues/356#issuecomment-1596196064) from @smaug----, and note that the post has received support emojis a developer too.

For sake of making progress here and avoiding the old spec discussion again without supporting data, I will close this issue and stick to the conclusion (https://github.com/w3c/pointerevents/issues/508#issuecomment-2258736459) that exposing the change seems to be the only way to know about possible compat problems.

**Comment by masayuki-nakano** [comment]
Okay, then, I'll update the tricky code before enabling the new behavior on Firefox.

**Comment by masayuki-nakano** [comment]
@mustaqahmed Gecko/Firefox started dispatching `click` event on the preceding `pointerup` event target in nightly builds. However, [we got a compat issue for a single tap](https://bugzilla.mozilla.org/show_bug.cgi?id=1966569). The case is, there is a popup look list whose list items are `<a>` elements and the list is set to `style.display="none"` when `mousedown`. Therefore, the `click` target becomes an element under the `<a>` when user taps an item since the element at `pointerdown` implicitly captures the pointer and `pointerup` will be fired on the same target even after it's hidden. ([Here is a testcase.](https://jsfiddle.net/d_toybox/ctnvpxm3/3/))

However, this is not reproducible with Chrome because Chrome does not behave so for the `click` caused by a touch. So, Chrome does not treat the `userEvent` as the preceding `pointerup`. Could you clarify that it's a bug of Chrome or `userEvent` is the `touchend` or `mouseup` for Chrome?

**Comment by masayuki-nakano** [comment]
(For now, Firefox will stop using the standardized target of the `click` caused by the compatibility mouse events of a touch.)

**Comment by mustaqahmed** [comment]
We didn't get a chance to investigate the use of `userEvent` for Chrome's touch click dispatch.  Touch clicks are fired from internal gesture tap events, and the state propagation between `PointerEventManager` and `GestureManager` is [buggy](https://issues.chromium.org/issues?q=componentid:1456874%20status:open%20touch%20pointer%20click%20capture).

I just filed [a new Chrome bug](https://issues.chromium.org/419497264) to track this, thanks for the report and also for the minimal repro.
